### ISSUE TITLE
Enforce max scan concurrency on engagement

### ISSUE THREAD

**Issue by EndPositive** [secureCodeBox/secureCodeBox#730]
## ➹ New Feature implementation request
Currently there is no limit on the amount of scans that may be triggered within an engagement. The Cascading Scan hook may create an unlimited amount scans, which are then gladly processed by the secureCodeBox operator and scheduled by Kubernetes. There are (at least) two problems with this approach:

1. The scanned endpoint may be bombarded by scans.
2. A single engagement may use up all Kubernetes resources.

### Describe the solution you'd like
Scans are labelled with a `pool` field. Before starting a scan, the Operator verifies the amount of Scans already running within that pool of scans. If there are already too many running, the Scan is temporarily evicted until there is space in the pool.

I'm not sure yet if we'd like to define this max concurrency per pool or as a global defined in operator.

### Describe alternatives you've considered
Problem 2 may be solved by deploying secureCodeBox scanners/hooks in different namespace per user and then using resource quotas on that namespace. Not ideal: would require lots extra admin work to maintain deployments across namespaces.

### Additional context
GitLab CI/CD defines job [concurrency on runner-basis](https://github.com/zaproxy/zap-extensions/pull/3240).

**Comment by rseedorff** [comment]
Hi @EndPositive,

thx for raising this issue 👍🏻 

Have you considered to use the Kubernetes namespace quota definitions to reduce the amount of concurrent scans? We are using this approach to slow down the concurrent scans based on the size of the Kubernetes cluster successfully.

**Comment by EndPositive** [comment]
> Have you considered to use the Kubernetes namespace quota definitions to reduce the amount of concurrent scans?

Yep! We're running all scanners with defined resource requests such that the cluster itself is never bombarded.

It is however difficult to slow down based on the engagement requirements (Issue 1 would not be solved by your solution). If we're scaling up our cluster, it would still be possible to launch 20 different scanners on each detected port/service, potentially DOSsing their servers. Ideally, we want do define per engagement how heavy it may be.

The intensity labels on cascading rules are not a solution as we still want to run intense cascading rules, just not at the same time as many other scanners.

**Comment by J12934** [comment]
Related: It is possible to define extended resources for for resource limits: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources
This should probably be possible to use to fit most of the requirements of this issue.

**Comment by sanmai-NL** [comment]
I think rate limiting, parallellism and resource pooling are complementary, so also partly distinct. 
A single scanner could still require a rate limit, even as it uses minimal resources.
Even if there are ample resources, you may still want to limit parallellism for debugging or experimental purposes.
If an rate limit is set that exceeds resource saturation, it should still allow for scaling up if more resources are added or become available suddenly within a long-running engagement.

**Comment by Weltraumschaf** [comment]
We have the point of resource optimization on the road map -> backlog

**Comment by J12934** [comment]
This can now be done with https://kueue.sigs.k8s.io/
We added a feature a coule of releases back to allow ScanTypes, ParseDefinitions and ScanCompletionHooks to start as suspended k8s jobs and then gets unsuspended by kueue based on the rules defined by the kueue CRDs
